### PR TITLE
Use map instead of vector for editableVars and other things

### PR DESCRIPTION
**Pull request by sway**
- Use map instead of vector to avoid iterating through all the vars for parameter updates
- Send all updated parameters back in the POST response (like GraphQL mutation results) to avoid having to re-fetch parameters after change from the FE
- Dynamically allocate the JSON object depending on the request method and number of parameters (also calculate in the JSON_STRING as otherwise the objects sometimes ends up being too small)

### CODE REVIEW

File: src/EmbeddedWebserver.h [hunk 1]
```diff
@@ -318,78 +340,64 @@ void serverSetup() {
...
             }
 
-            request->send(200, "text/html", "Parameters saved");
+            String paramsJson;
+            serializeJson(doc, paramsJson);
+            request->send(200, "application/json", paramsJson);
 
             // Write to EEPROM
-            if (writeToEeprom) {
+            if (false) {
```
Comment (sway) [inline src/EmbeddedWebserver.h:400]: Just leftover from testing 😓

File: src/EmbeddedWebserver.h [hunk 2]
```diff
@@ -318,78 +340,64 @@ void serverSetup() {
...
-            /*String m = "Got ";
-            m += params;
-            m += " request parameters: <br />";
-            */
+        // Determine the size of the document to allocate based on the number
+        // of parameters
+        // GET = either
+        int requestParams = request->params();
+        int docLength = EDITABLE_VARS_LEN;
+        if(request->method() == HTTP_GET) {
```
[the remark this answers is not in the document]
Comment (kjyv) [reply src/EmbeddedWebserver.h:353]: That's why it was numerical

File: src/EmbeddedWebserver.h [hunk 3]
```diff
@@ -399,85 +407,62 @@ void serverSetup() {
 
             // Write the new values to MQTT
             writeSysParamsToMQTT();
-
-        } else if (request->method() == 1) {  //WebRequestMethod enum -> HTTP_GET
-            //return all params as json
-            DynamicJsonDocument doc(JSON_ARRAY_SIZE(1) + JSON_OBJECT_SIZE(9+1) * EDITABLE_VARS_LEN);
-
-            //get parameter id from frst parameter, e.g. /parameters?param=PID_ON
+        } else if (request->method() == HTTP_GET) {
```
Comment (kjyv) [inline src/EmbeddedWebserver.h:410]: same as above, HTTP_GET is probably not correct (unless it resolves to 1 now)

File: src/EmbeddedWebserver.h [hunk 4]
```diff
@@ -399,85 +407,62 @@ void serverSetup() {
...
+        String paramsJson;
+        serializeJson(doc, paramsJson);
+        request->send(200, "application/json", paramsJson);
     });
 
     server.on("/parameterHelp", HTTP_GET, [](AsyncWebServerRequest *request) {
         DynamicJsonDocument doc(1024);
 
-        AsyncWebParameter* p = request->getParam(0);
+        AsyncWebParameter *p = request->getParam(0);
```
Comment (kjyv) [inline src/EmbeddedWebserver.h:449]: why this change? Maybe it's a matter of taste but I read it as "the type of the dereferenced pointer variable p is set to  AsyncWebParameter" so having it next to p makes more sense to me :)

File: src/EmbeddedWebserver.h [hunk 5]
```diff
@@ -492,25 +477,23 @@ void serverSetup() {
...
 
-        //go through history values backwards starting from currentIndex and wrap around beginning
-        //to include valueCount many values
-        for (int i=mod(historyCurrentIndex-historyValueCount, HISTORY_LENGTH);
-                 i!=mod(historyCurrentIndex, HISTORY_LENGTH);
-                 i=mod(i+1, HISTORY_LENGTH))
-        {
+        // go through history values backwards starting from currentIndex and
+        // wrap around beginning to include valueCount many values
+        for (int i = mod(historyCurrentIndex - historyValueCount, HISTORY_LENGTH); i != mod(historyCurrentIndex, HISTORY_LENGTH); i = mod(i + 1, HISTORY_LENGTH)) {
```
Comment (kjyv) [inline src/EmbeddedWebserver.h:496]: line is too long now, was better before

File: src/main.cpp [hunk 6]
```diff
@@ -328,7 +329,8 @@ enum SectionNames {
     sOtherSection
 };
 
-std::vector<editable_t> editableVars = {};
+//std::vector<editable_t> editableVars = {};
```
Comment (kjyv) [inline src/main.cpp:332]: line can be removed now

File: src/main.cpp [hunk 7]
```diff
@@ -1397,82 +1399,57 @@ void websiteSetup() {
 const char sysVersion[] = (STR(FW_VERSION) "." STR(FW_SUBVERSION) "." STR(FW_HOTFIX) " " FW_BRANCH);
 
 void setup() {
-    editableVars =  {
-        //#1
-        {F("PID_ON"), "Enable PID Controller", false, "", kUInt8, 0, []{ return true; }, 0, 1, (void *)&pidON},
+    //#1
+    editableVars["PID_ON"] = {F("PID_ON"), "Enable PID Controller", false, "", kUInt8, 0, 1, []{ return true; }, 0, 1, (void *)&pidON};
```
Comment (kjyv) [inline src/main.cpp:1403]: string duplication is not so nice here, do we still need the name again in the map entry?

File: src/EmbeddedWebserver.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (sway) [reply src/EmbeddedWebserver.h:353]: In the IDE it shows that the return type is `WebRequestMethodComposite`, which is the same as what's in the `on` handler signature, where we're already using the enum, so it should be fine. `HTTP_GET` and `HTTP_POST` resolve to 1 and 2, respectively.
Maybe something that was changed in a later version of ESPAsyncWebServer?

File: src/EmbeddedWebserver.h (hunk 4 above)
Comment (sway) [reply src/EmbeddedWebserver.h:449]: Fine with me, this got auto-formatted by my IDE so happy to keep the `*` next to the type

File: src/main.cpp (hunk 7 above)
Comment (sway) [reply src/main.cpp:1403]: Good point, we can slim it down and use the key instead 👍🏻

File: src/EmbeddedWebserver.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (kjyv) [reply src/EmbeddedWebserver.h:353]: WebRequestMethodComposite is just an alias for unit8_t in ESPAsyncWebServer (of which there is no newer versions since a long time, I have the same), but it's compared to HTTPMethod which is part of ESP8266WebServer.h or for ESP32 a different file. Maybe these differ then, so the code is not ESP8266 compatible. HTTP_POST is 3 in this case.

[the remark this answers is not in the document]
Comment (kjyv) [reply src/EmbeddedWebserver.h:353]: The only problem really is that HTTP_GET and HTTP_POST are a different type than what on() and request->method() use. It should be WebRequestMethod.HTTP_GET instead but we've have that disable with #define WEBSERVER_H for some reason

[the remark this answers is not in the document]
Comment (sway) [reply src/EmbeddedWebserver.h:353]: Yeah, you're right, I think my IDE played a trick on me (or my eyes did). Will revert it to the explicit number comparison (and add another comment so that no one else tries to be a smartass like I was:))

File: src/EmbeddedWebserver.h [hunk 8]
```diff
@@ -318,19 +258,32 @@ void serverSetup() {
         request->redirect("/");
     });
 
-    server.on("/parameters", HTTP_GET | HTTP_POST, [](AsyncWebServerRequest *request) {
-        //stop writing to heater in ISR method (digitalWrite) as it causes crashes when called at the same time as flash is read or written
+    server.on("/parameters", HTTP_GET | HTTP_POST,[](AsyncWebServerRequest *request) {
```
Comment (kjyv) [inline src/EmbeddedWebserver.h:261]:
```suggestion
    server.on("/parameters", HTTP_GET | HTTP_POST, [](AsyncWebServerRequest *request) {
```

File: src/EmbeddedWebserver.h [hunk 9]
```diff
@@ -318,19 +258,32 @@ void serverSetup() {
...
-            /*String m = "Got ";
-            m += params;
-            m += " request parameters: <br />";
-            */
+        // Determine the size of the document to allocate based on the number
+        // of parameters
+        // GET = either
+        int requestParams = request->params();
+        int docLength = EDITABLE_VARS_LEN;
+        if(request->method() == 1) {
```
Comment (kjyv) [inline src/EmbeddedWebserver.h:271]:
```suggestion
        if (request->method() == 1) {
```

File: src/EmbeddedWebserver.h [hunk 10]
```diff
@@ -318,19 +258,32 @@ void serverSetup() {
...
+            if(requestParams > 0) {
```
Comment (kjyv) [inline src/EmbeddedWebserver.h:272]:
```suggestion
            if (requestParams > 0) {
```

File: src/EmbeddedWebserver.h [hunk 11]
```diff
@@ -318,19 +258,32 @@ void serverSetup() {
...
+            }
+        } else if (request->method() == 2) {
+            docLength = std::min(requestParams, EDITABLE_VARS_LEN);
+        }
+
+        DynamicJsonDocument doc(JSON_STRING_SIZE(262) + JSON_ARRAY_SIZE(docLength) +
+                                JSON_OBJECT_SIZE(9 + 1) * docLength);
 
-            for (int i = 0 ; i < params; i++) {
+        if (request->method() == 2) {
```
Comment (kjyv) [inline src/EmbeddedWebserver.h:282]:
```suggestion
        if (request->method() == 2) {   // method() returns values from WebRequestMethod enum -> 2 == HTTP_POST
```

File: src/EmbeddedWebserver.h [hunk 12]
```diff
@@ -406,56 +332,36 @@ void serverSetup() {
...
-
-            for (editable_t e : editableVars) {
-                if (!paramId.isEmpty()) {
-                    //we have a parameter to select a single var,
-                    //skip vars until we find the one for the id in the request parameter
-                    if (paramId != e.templateString) {
-                        continue;
-                    }
-                }
+            String paramId =
```
Comment (kjyv) [inline src/EmbeddedWebserver.h:335]: should be in one line
